File: src/Commands/all/PageRef.ts
import VuePagRef from "../Components/VuePagRef.vue";

export class PageRef {
	public static symbol: string = '..';
	public static commandName: string = 'pageref';

	public pageRef: number;
	public allPageRefs: Array<[string | null, number, number?]>;
	public pagesString: string;
	public resourceAlias?: string;

	constructor(pageRef: string) {
		let alias: string | undefined;
		let pageString = pageRef || '';

		// Support alias separator with ':' or ';' (first occurrence)
		const idxColon = pageRef.indexOf(':');
		const idxSemi = pageRef.indexOf(';');
		const idx = [idxColon, idxSemi]
			.filter((i) => i !== -1)
			.sort((a, b) => a - b)[0];

		if (idx !== undefined) {
			alias = pageRef.slice(0, idx).trim();
			pageString = pageRef.slice(idx + 1).trim();
		}

		this.pagesString = pageString;
		this.resourceAlias = alias;
		this.allPageRefs = this.parsePageRefs(pageString?.toString() || '', alias);
		this.pageRef = (this.allPageRefs[0]?.[1] ?? 0);
	}

	private parsePageRefs(pageRefString: string, defaultAlias?: string): Array<[string | null, number, number?]> {
		const result: Array<[string | null, number, number?]> = [];

		// Tokenize by ',' and ';' while allowing alias overrides per token.
		const tokens = pageRefString
			.split(/[;,]/)
			.map((t) => t.trim())
			.filter((t) => t.length > 0);

		let currentAlias: string | null = defaultAlias ?? null;

		for (const token of tokens) {
			// If token has an explicit alias using ':' then it sets/overrides current alias
			let work = token;
			let aliasForToken: string | null = null;
			const pos = token.indexOf(':');
			if (pos !== -1) {
				aliasForToken = token.slice(0, pos).trim() || null;
				work = token.slice(pos + 1).trim();
				// Persist alias for following tokens until changed
				currentAlias = aliasForToken;
			}

			// If token looks like a bare alias (no digits and no colon), treat it as a context switch
			if (pos === -1 && !/[0-9]/.test(work)) {
				currentAlias = work || null;
				continue;
			}

			if (!work) continue;

			// Parse either range "a-b" or single number, with optional scroll percentage "num%scroll"
			let scrollPercent: number | undefined = undefined;
			const percentPos = work.indexOf('%');
			if (percentPos !== -1) {
				const scrollStr = work.slice(percentPos + 1).trim();
				const scrollNum = parseInt(scrollStr, 10);
				if (!isNaN(scrollNum)) {
					scrollPercent = scrollNum;
				}
				work = work.slice(0, percentPos).trim();
			}

			if (work.includes('-')) {
				const [startStr, endStr] = work.split('-').map((s) => s.trim());
				const start = parseInt(startStr, 10);
				const end = parseInt(endStr, 10);
				if (!isNaN(start) && !isNaN(end)) {
					for (let i = start; i <= end; i++) {
						result.push([aliasForToken ?? currentAlias ?? null, i, scrollPercent]);
					}
				}
			} else {
				const num = parseInt(work, 10);
				if (!isNaN(num)) {
					result.push([aliasForToken ?? currentAlias ?? null, num, scrollPercent]);
				}
			}
		}

		return result;
	}

	public toJson(): IPageRef {
		return {
			name: PageRef.commandName,
			vueComponent: VuePagRef,
			ref: this.pageRef,
			allRefs: this.allPageRefs,
			pagesString: this.pagesString,
			resourceAlias: this.resourceAlias
		};
	}
}

export interface IPageRef {
	name: string;
	vueComponent: any;
	ref: number;
	allRefs: Array<[string | null, number, number?]>;
	pagesString: string;
	resourceAlias?: string;
}

File: service/flashcard_parser/commands/page_ref.py
class PageRef:
    symbol = ".."
    command_name = "pageref"

    def __init__(self, page_ref: str):
        alias = None
        page_string = page_ref
        idx = page_ref.find(":")
        if idx != -1:
            alias = page_ref[:idx].strip()
            page_string = page_ref[idx + 1 :].strip()
        self.pages_string = page_string
        self.resource_alias = alias
        self.all_page_refs = self._parse_page_refs(page_string or "0")
        self.page_ref = self.all_page_refs[0] if self.all_page_refs else 0

    def _parse_page_refs(self, page_ref_string: str):
        result = []
        parts = [p.strip() for p in page_ref_string.split(",")]
        for part in parts:
            if "-" in part:
                start, end = part.split("-")
                try:
                    start_i = int(start.strip())
                    end_i = int(end.strip())
                    for i in range(start_i, end_i + 1):
                        result.append(i)
                except ValueError:
                    pass
            else:
                try:
                    result.append(int(part))
                except ValueError:
                    pass
        return result

    def to_json(self):
        return {
            "name": "PageRef",
            "ref": self.page_ref,
            "allRefs": self.all_page_refs,
            "pagesString": self.pages_string,
            "resourceAlias": self.resource_alias,
        }
